src/main/scala/nl/knaw/dans/easy/pid/Command.scala

```diff
@@ -30,27 +31,38 @@ object Command extends App with DebugEnhancedLogging {
     verify()
   }
   val app = new PidGeneratorApp(new ApplicationWiring(configuration))
-  val result: Try[FeedBackMessage] = commandLine.subcommand match {
-    case Some(generate @ commandLine.generate) => generate.pidType() match {
-      case "doi" => app.generate(DOI)
-      case "urn" => app.generate(URN)
-    }
-    case Some(_ @ commandLine.runService) => runAsService()
-    case _ => Failure(new IllegalArgumentException(s"Unknown command: ${ commandLine.subcommand }"))
-  }
 
-  result.doIfSuccess(msg => println(s"OK: $msg"))
+  managed(app)
+    .acquireAndGet(app => {
+      for {
+        _ <- app.init()
+        msg <- runSubcommand(app)
+      } yield msg
+    })
+    .doIfSuccess(msg => println(s"OK: $msg"))
     .doIfFailure { case e => logger.error(e.getMessage, e) }
     .doIfFailure { case NonFatal(e) => println(s"FAILED: ${ e.getMessage }") }
 
-  app.destroy()
+  private def runSubcommand(app: PidGeneratorApp): Try[FeedBackMessage] = {
+    commandLine.subcommand
+      .collect {
+        case generate @ commandLine.generate => generate.pidType() match {
+          case "doi" => app.generate(DOI)
+          case "urn" => app.generate(URN)
+          case t => Failure(new IllegalArgumentException(s"Unknown PID type: $t"))
+        }
+        case commandLine.runService => runAsService(app)
+      }
+      .getOrElse(Failure(new IllegalArgumentException(s"Unknown command: ${ commandLine.subcommand }")))
+  }
 
-  private def runAsService(): Try[FeedBackMessage] = Try {
+  private def runAsService(app: PidGeneratorApp): Try[FeedBackMessage] = Try {
     val service = new PidGeneratorService(configuration.properties.getInt("pid-generator.daemon.http.port"), app)
     Runtime.getRuntime.addShutdownHook(new Thread("service-shutdown") {
       override def run(): Unit = {
         service.stop()
         service.destroy()
+        app.close()
```

Comment:
if we run the service this `run` method is the last thing it does when closing. It will not return to the function that called `runAsService`, since we terminate the application unexpectedly. Therefore we need to do a manual `app.close()` here as well.